### PR TITLE
Test: start a test suite for migration utils functions

### LINKED ISSUE #8042: Add record count idempotency tests for `run_key_migration_functions`
Add Django unit tests that verify the `run_key_migration_functions` management command only creates missing migration records on the first run, and does not create duplicate records when run again.

The test should capture relevant database record counts before the command runs, after the first run, and after a second run. The first run is expected to create or backfill records. The second run must leave all tracked record counts unchanged.

Prior to running the test, we'll need to construct database records similar to what Specify 6 does, in order to better catch actual edge cases. We should also add record changes that could occur with normal Specify 7 usage both before the first run, as well as between the first and second run.

## Background

`run_key_migration_functions` re-runs several important data migration helpers, including collection object type setup, schema config setup, permissions initialization, business rule initialization, app resource directory setup, tectonic rank setup, and misc backfills.  This was done to handle bringing in new records from Specify 6, which required all the key migration functions to handle being run multiple times, rather than just once.

Because this command may be run more than once in deployed environments, we need test coverage proving that it is idempotent with respect to record creation. Running the command a second time should not create duplicate records.

## Implementation Plan

1. Add a new Django test module, likely:
   `specifyweb/specify/tests/test_run_key_migration_functions.py`

2. Use the existing `ApiTests` fixture base from `specifyweb/specify/tests/test_api.py`.
   This already creates a realistic institution/division/discipline/collection/user setup.

3. Add helper functions:
   - `record_counts()` to snapshot counts for tracked models.
   - `count_diff(before, after)` to return only models whose counts changed.

4. Track models touched by the command, especially models where first-run creation is expected:
   - `Collectionobjecttype`
   - `Picklist`
   - `Picklistitem`
   - `Splocalecontainer`
   - `Splocalecontaineritem`
   - `Splocaleitemstr`
   - `UniquenessRule`
   - `UniquenessRuleField`
   - `LibraryRole`
   - `LibraryRolePolicy`
   - `Role`
   - `RolePolicy`
   - `UserRole`
   - `UserPolicy`
   - `Spappresourcedir`
   - `Tectonicunittreedef`
   - `Tectonicunittreedefitem`
   - `Tectonicunit`

5. In the test:
   - Snapshot counts before running the command.
   - Run `call_command("run_key_migration_functions", stdout=StringIO())`.
   - Snapshot counts after the first run.
   - Assert that at least one expected model count increased.
   - Run the command a second time.
   - Snapshot counts again.
   - Assert that the second-run diff is empty.

### LINKED ISSUE #8049: Add unit test coverage for `run_key_migration_functions`
The `run_key_migration_functions` management command re-runs several important data migration sections, but it currently has no focused unit test coverage. Add tests that verify the command dispatches the correct sections, preserves execution order, passes `apps` and optional verbose output correctly, and covers the local database logic implemented directly in the command module.

Add a new back-end tests for each of the key migration functions:

- `apply_patches`
- `fix_cots`
- `fix_permissions`
- `fix_business_rules`
- `fix_schema_config`
- `fix_app_resource_dirs`
- `fix_tectonic_ranks`
- `fix_misc`

We can try writing tests for all the specific migration functions in all of these sections.  That would be a lot, so maybe we could pick some select functions of interest to write unit tests for.

### LINKED ISSUE #8224: Add unit tests for migration code

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #8224 
Fixes #8049 
Fixes #8042 

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add relevant issue to release milestone
- [ ] Add pr to documentation list
- [ ] Add automated tests
- [ ] Add a reverse migration if a migration is present in the PR
- [ ] Add migration function to https://github.com/specify/specify7/blob/ea04665987e1b8a1b76955c7b7f702b7bf701b47/specifyweb/specify/management/commands/run_key_migration_functions.py#L50

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved migration reliability and idempotency across several data/schema update paths.
  * Fixed duplicate-record handling and ensured repeated migration runs no longer create extra entries.
  * Corrected handling of default values, hidden fields, and related schema updates for several core data types.

* **Tests**
  * Added broad coverage for migration commands and helper workflows.
  * Expanded database-backed tests for schema updates, deduplication, and rollback behavior.

### CODE REVIEW

**Comment by CarolineDenis** [comment]
@coderabbit review

**Comment by CarolineDenis** [comment]
@coderabbit review

**Comment by CarolineDenis** [comment]
@coderabbit review

**Comment by CarolineDenis** [comment]
@coderabbit review

**Comment by CarolineDenis** [comment]
@coderabbit full review

File: specifyweb/specify/management/commands/tests/cots_tests.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+from unittest.mock import Mock, sentinel
+from django.apps import apps as django_apps
+from django.test import SimpleTestCase
+from specifyweb.specify import models
+from specifyweb.specify.tests.test_api import ApiTests
+from specifyweb.specify.management.commands import run_key_migration_functions as rkm
+from specifyweb.specify.management.commands.tests.test_migration_base import MigrationCommandTestCase
+
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/cots_tests.py]: [8dbe71b](https://github.com/specify/specify7/pull/8225/commits/8dbe71b10c1fc16116ba6842a8ea059602432c76)

File: specifyweb/specify/management/commands/tests/schema_config_tests.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+            "update_co_children_fields",
+            "remove_collectionobject_parentco",
+            "add_quantities_gift",
+            "update_paleo_desc",
+            "update_accession_date_fields",
+            "componets_schema_config_migrations",
+            "create_discipline_type_picklist",
+            "update_discipline_type_splocalecontaineritem",
+            "deduplicate_schema_config_orm",
+        ]
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/schema_config_tests.py:52]: [06e8f30](https://github.com/specify/specify7/pull/8225/commits/06e8f30b95964f869e3944ba07f9a1e346ce5b80)

File: specifyweb/specify/management/commands/tests/schema_config_tests.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                },
+            ],
+        )
+
+        keeper.refresh_from_db()
+        self.assertEqual(created_count, 1)
+        self.assertEqual(keeper.text, "Updated Name")
+        self.assertFalse(
+            models.Splocaleitemstr.objects.filter(id=duplicate.id).exists()
+        )
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/schema_config_tests.py:149]: [f7bb933](https://github.com/specify/specify7/pull/8225/commits/f7bb933019b17e394e5991bdb186a43bad45a1b6)

File: specifyweb/specify/management/commands/tests/schema_config_tests.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        self.assertEqual(duplicate_name.itemname_id, keeper.id)
+        self.assertEqual(duplicate_desc.itemdesc_id, keeper.id)
+        self.assertFalse(
+            models.Splocaleitemstr.objects.filter(
+                id=duplicate_conflicting_name.id,
+            ).exists()
+        )
+        self.assertEqual(
+            set(keeper.names.values_list("language", "text")),
+            {("en", "Keeper Name"), ("es", "Duplicate Name ES")},
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/schema_config_tests.py]: [ec885eb](https://github.com/specify/specify7/pull/8225/commits/ec885eb647426d406ceacc4b4343b620119c6004)

File: specifyweb/specify/management/commands/tests/tectonic_tests.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+from specifyweb.specify.management.commands import run_key_migration_functions as rkm
+from django.apps import apps as django_apps
+
+class TectonicDatabaseTests(TestCase):
+    def test_create_default_tectonic_ranks_creates_chain_and_assigns_discipline(self):
+        self.discipline.tectonicunittreedef = None
+        self.discipline.save()
+        models.Tectonicunittreedef.objects.filter(
+            discipline=self.discipline,
+        ).delete()
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/tectonic_tests.py]: [88d3588](https://github.com/specify/specify7/pull/8225/commits/88d35882c4a0e6a535b9bda45d40a73233ad7b16)

File: specifyweb/specify/migration_utils/tests/test_bulk_create.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        self.assertTrue(mock_manager.bulk_create.called)
+
+        # -----------------------
+        # Inspect bulk_create payload
+        # -----------------------
+        args, _kwargs = mock_manager.bulk_create.call_args
+        created_objects = args[0]
+
+        self.assertEqual(len(created_objects), 2)
+
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/migration_utils/tests/test_bulk_create.py]: [9d48f36](https://github.com/specify/specify7/pull/8225/commits/9d48f36d45aceac897ba067f49e50b4e73675d2c)

**Comment by CarolineDenis** [comment]
@coderabbit full review

File: specifyweb/specify/management/commands/tests/test_run_key_migration_functions_order.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            stack.enter_context(patch.object(rkm, "apps", sentinel.apps))
+            self._patch_recorders(stack, patch_targets, calls)
+
+            section(stdout)
+
+        self.assertEqual(calls, [(name, sentinel.apps) for name in expected_names])
+        self.assertEqual(
+            stdout.call_args_list,
+            [call(f"Running {name}...") for name in expected_names],
+        )
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/test_run_key_migration_functions_order.py]: [95aaefe](https://github.com/specify/specify7/pull/8225/commits/95aaefe7b3e104402b866688bb0554dcdc2d3944)

File: specifyweb/specify/migration_utils/tests/test_helper_0042_discipline_type_picklist.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        helper_0042_discipline_type_picklist.update_discipline_type_splocalecontaineritem(
+            apps
+        )
+
+        helper_0042_discipline_type_picklist.revert_discipline_type_splocalecontaineritem(
+            apps
+        )
+
+        qs = self.collection.__class__.objects.all()
+        self.assertTrue(qs.exists())
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/migration_utils/tests/test_helper_0042_discipline_type_picklist.py]: [68e7fb3](https://github.com/specify/specify7/pull/8225/commits/68e7fb3a67fca1f87e6bba5132c73b2e300ed5a5)

File: specifyweb/specify/management/commands/tests/test_run_key_migration_functions.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+            "fix_schema_config",
+            "fix_app_resource_dirs",
+            "fix_permissions",
+            "fix_business_rules",
+            "fix_tectonic_ranks",
+            "fix_misc",
+        )
+
+        after_first_run = record_counts()
+        first_run_diff = count_diff(before_first_run, after_first_run)
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/test_run_key_migration_functions.py]: [118e4b3](https://github.com/specify/specify7/pull/8225/commits/118e4b384e518b39c015b2c8ac61775a466a69cb)

File: specifyweb/specify/management/commands/tests/schema_config_tests.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        self.assertEqual(fake_apps.model_request, ("specify", "Discipline"))
+        self.assertEqual(
+            calls,
+            [(name, fake_apps) for name in names[:-1]]
+            + [
+                ("apply_schema_defaults_task.apply", [discipline_1.id]),
+                ("apply_schema_defaults_task.apply", [discipline_2.id]),
+                (names[-1], fake_apps),
+            ],
+        )
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/schema_config_tests.py:80]: [49a803d](https://github.com/specify/specify7/pull/8225/commits/49a803d9e735688b60b2eb3259038f956c3a37b2)

File: specifyweb/specify/management/commands/tests/business_rules_tests.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        apply_rules.assert_called_once_with(
+            discipline_without_constraint,
+            registry=fake_apps,
+        )
+        self.assertIs(fake_uniqueness_rule_manager.last_is_database_constraint, True)
+
+
+class BusinessRulesDatabaseTests(TestCase):
+    def test_catnum_rule_editable_only_updates_matching_catalog_number_rule(self):
+        discipline = Discipline.objects.create(name="Test Discipline")
```
[the remark this answers is not in the document]
Comment (CarolineDenis) [reply specifyweb/specify/management/commands/tests/business_rules_tests.py]: [d9f93ac](https://github.com/specify/specify7/pull/8225/commits/d9f93ac0ded821fc95002ea2ba43a6f22f9b133a)

File: specifyweb/specify/management/commands/tests/business_rules_tests.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        fake_apps = FakeApps()
+
+        with patch.object(rkm, "apply_default_uniqueness_rules") as apply_rules:
+            rkm.apply_default_uniqueness_rules_to_disciplines(fake_apps)
+
+        apply_rules.assert_called_once_with(
+            discipline_without_constraint,
+            registry=fake_apps,
+        )
+        self.assertIs(fake_uniqueness_rule_manager.last_is_database_constraint, True)
```
Comment (melton-jason) [inline specifyweb/specify/management/commands/tests/business_rules_tests.py:70]: (A note on how we can expand this test)

Currently this tests for the control flow of `apply_default_uniqueness_rules_to_disciplines -> apply_default_uniqueness_rules` (i.e., that the former calls the latter).
As far as I know, we don't have tests for the overall functionality of `apply_default_uniqueness_rules`--that is tests that make sure that:
1. Rules are created in Disciplines 
2. Duplicate rules are not created in Disciplines which already have the rules

And there currently are not tests for `apply_default_uniqueness_rules_to_disciplines` within run_key_migration_functions that makes sure `apply_default_uniqueness_rules` is only applied to Disciplines that don't contain Uniqueness Rules.

File: specifyweb/specify/management/commands/tests/tectonic_tests.py
Comment (melton-jason) [inline specifyweb/specify/management/commands/tests/tectonic_tests.py:1]: (A very minor note on the location of these tests)

Can these tests be merged/added with the ones in `migration_utls/tests/test_tectonic_ranks.py`? I think the only command that references Tectonic Unit is `run_key_migration_functions`, and the functions tested in this file and the migration_utils file should be the same.

File: specifyweb/specify/migration_utils/tests/test_bulk_create.py
Comment (melton-jason) [inline specifyweb/specify/migration_utils/tests/test_bulk_create.py:1]: (A note about mocks and how we can make these tests more representative of user use and data)

When it comes to the Schema Configuration functions being executed in migrations and/or run_key_migration_functions, I think we should heavily prefer using the database models and ORM rather than mocks. This would largely be due to the complexity of the involved operation(s) and models, and the customizability of the Schema Configuration as a whole. 
Mocks require imposing some control over the environment and "migration internals" that isn't present in a non-test situation (i.e., production. Sort of like how the frontend tests can run isolated from the backend but the responses have to be mocked in good faith to the corresponding requests: the response data is entirely controlled by the developer and in something like an isolated sandbox, which is not the case in prodcution or development environments). 
A more blackbox and result-focused approach (instead of the mock approach which requires expectations of how the functions are internally implemented) would mean that changes to the function internals that are still equivalent would not result in failing tests in the future. 

All this to say I think it would be great to migrate these Schema Configuration tests to leverage the database and non-mocked functionality as much as possible (mocking something like stdout in run_key_migraiton_functions would still be a perfect use case for a mock)!

**Review by melton-jason — APPROVED** [review]
Thank you for writing these tests! 🥳 

I've modified some of the tests and fixed the test that was failing. See the recent commits [`db45d99..9d65f24` (this PR)](https://github.com/specify/specify7/pull/8225/changes/db45d99033d153d080e1a78b6113b7dc24986fe8..9d65f24c067ae86c63f172cea5ddd44de5e0d8ba).

I left some notes about how some tests can be expanded on in the future. 
Particularly, I think there are some tests that are currently utilizing mocks that can be re-written to be a little less reliant on mocks, or at least new tests can be written that mirror the behavior in production. 
While significantly slower than mocks when it comes to testing with the Django ORM/database, some of the potential downsides of mocks are:
1. With the "white-box" nature of mocking the internals of functions and classes (e.g., mocking some function B that function A calls when testing A), the tests can become very fragile to internal changes of the functions they're testing
  - If we modify some function being tested to have the same output, tests relying on mocks can fail because the mocks within the tests are relying the function to be implemented a certain way
  - Traditional "black-box" testing is only concerned with the output of the function given some input: the entire implementation of the function can change and the tests will be unaffected as long as the result/output remains the same
2. It is very hard to simulate _all_ aspects of a particular ORM/database operation. Things like database connectivity, Django signals, database locks, custom model methods, database constraints, etc. can all play a part in a production environment that are difficult to model accurately with mocks
  - Even if the mocks are modeled accurately, the mocked models would have to change with the production system they're modeling. Otherwise, the tests can continue to pass or exhibit undefined behavior while being divergent from how the operation is supposed to work in production environments

Not to say we should _never_ use mocks! I think mocks are great for things like `stdout` when working with `run_key_migration_functions`. 
In that case, the complete side effects of mocking stdout is well-known (text won't be displayed to the screen/terminal), and we can still treat a mocked stdout as a "black-box" (we don't care about any mutations that the original stdout would've performed, so we only need to care about how and what stdout was interacted with. For the stdout case, the side effects and non-mocked results are always relatively static and trivial to determine).

Things like testing default COT creation and updates, default uniqueness rules and schema configuration changes and deduplication could likely benefit in the long run with tests that leverage the database an non-mocked Django apps/models. 

Those tests can definitely be part of a separate PR, and I can write up Issues for them if needed! 
Or let me know if there's some aspect of the current implementation of tests that I may be overlooking! Was not too familiar with mocks before this PR-- I have spent some time going through Python's [unittest.mock docs](https://docs.python.org/3.12/library/unittest.mock.html#module-unittest.mock) and there is still a lot to learn 🎓.

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Speciforum**. There might be relevant details there:

https://speciforum.org/t/specify-7-12-0-7-release-announcement/3863/1
